### ISSUE TITLE
Can clients change their election_id across different WriteRequest messages?

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4runtime#121]
12. Write RPC

Should a server check the election_id value in every WriteRequest to
see if it matches what the same client has sent before?

Should a client ever be allowed to change its election_id in a single
session?

**Comment by jafingerhut** [comment]
Regarding my question "Should a client ever be allowed to change its election_id in a single session?" it appears that perhaps the answer is "yes", according to Section 5.3 "Rules for handling MasterArbitrationUpdate Messages ...", number 2. (b) iii.

"Otherwise, the P4Runtime server updates the `election_id` for this controller. If this makes the client the new master, the server will also accept the given `role.config` and follow the "mastership change rules" described in the [following section](#sec-master-change)."

It isn't obvious whether that is the _only_ way that a client's election_id is allowed to change.

**Comment by antoninbas** [comment]
The election_id value in each `WriteRequest` is verified to check that the client is authorized to perform the write:

> The device_id uniquely identifies the target P4 device. The role_id and election_id define the client role and election-id as described in the Master-Slave Arbitration and Controller Replication section. The server is expected to perform the following checks (in this order) before processing the updates list:
If device_id does not match any of the devices known to the P4Runtime server or if role_id does not match any of the roles for the device, the server must return a NOT_FOUND error.
If the client is not the master for (device_id, role_id) according to the election_id value, the server must return a PERMISSION_DENIED error.
If the Write is attempted before a ForwardingPipelineConfig has been set, the server must return a FAILED_PRECONDITION error.

When a `WriteRequest` is received, the server doesn't know which client the request originated from until it inspects the election_id. gRPC may not provide enough information to identify the client, and even if it did I would be in favor in baking this information in the protocol, rather than relying on the gRPC library.

We assume that the client doesn't lie. In practice clients will be identified using SSL/TLS. This is not mentioned in the spec yet, but I believe there is interest in including this information in a future minor release (1.1?).

Please fee free to add more information about election id change in the appropriate section.

**Comment by jafingerhut** [comment]
OK, using election_id to identify the client sounds reasonable, as well as trusting clients not to lie about their election_id.

What does this sentence mean, then?

"More specifically, the  P4Runtime server will know which stream corresponds to the master controller  for each pair of (`device_id`, `role_id`) at any point of time."

Also, when a stream is broken, which should trigger a change in master, isn't it true that the server _must_ rely upon getting some knowledge of which client the session was broken for?  Otherwise, how can it know that the master client session was broken, rather than a slave client session?

In the section titled "Rules for Handling `MasterArbitrationUpdate` Messages Received from Controllers", there are different rules for processing a MasterArbitrationMessage depending upon whether such a message is received from a client "for the first time" versus "from an already connected controller".  How does the server determine this, if not from some reliable way to identify a gRPC session?

**Comment by antoninbas** [comment]
I think there is some misunderstanding of what a gRPC stream is. They are fundamentally different from unary RPCs. Unary RPCs are stateless and don't let us identify the client. However, on the server side, a bi-directional stream RPC is actually a function call that never returns until the server decides to return (in synchronous mode, but it's logically equivalent in async mode). All the stream messages are received within the scope of this function call. Even when the stream breaks, the server does whatever it needs to do before exiting from that function call. Maybe looking at this code will clarify things: https://github.com/p4lang/PI/blob/master/proto/server/pi_server.cpp#L429. Basically the server keeps some connection state for the client: device_id, election_id, role_id, mastership status. When a Write unary RPC is received, the server can match the information in the `WriteRequest` against that connection state. That connection state can be updated throughout the lifetime of the client connection (logically its lifetime is the lifetime of the function call).

**Comment by jafingerhut** [comment]
Very likely I have been living with some misunderstandings about gRPC and its properties, so thanks for the explanation there.

So your earlier comment "gRPC may not provide enough information to identify the client, and even if it did I would be in favor in baking this information in the protocol, rather than relying on the gRPC library." applies for unary RPCs only.  For the bi-directional stream, the server process can identify that messages are received from the same client as earlier messages, because the function call that handles all received messages on this bi-directional stream is the same function that received and processed earlier messages from the same client?

If so, I cannot think of any remaining concerns I have about this topic.

**Comment by antoninbas** [comment]
Yes that's correct, my statement only applied to unary RPCs such as Write. We need to map each unary RPC to the connection state which is maintained by the stream RPC. The information required for the mapping has to be baked in the WriteRequest message.

**Comment by jafingerhut** [comment]
I will think about this a bit more and perhaps create a PR suggesting a few additional sentences that might help explain a couple of these issues, or at least remind me of them.  You can reserve judgement on how wise this is until after you've read the PR.  I know the goal is to rely on knowledge of Google Protobufs and gRPC and not explain them in this spec, but perhaps a few short comments may be reasonable.

I suppose this implies that every client and server must have at least one thread for handling the stream channel, and one more thread for handling other messages?  At least, I don't see a way to write a single-threaded implementation of P4Runtime with the stated approach for maintaining the stream channel.

**Comment by antoninbas** [comment]
Your statement is correct in synchronous mode, which I would recommend using on the server side. In asynchronous mode, one could get away with using a single thread servicing the gRPC completion queue if the objective was to minimize the number of threads. However, it takes a braver man to write some asynchronous gRPC streaming code :)

**Comment by jafingerhut** [comment]
Should ReadRequest messages contain a role_id and election_id field for the same reason that WriteRequest messages do?

**Comment by antoninbas** [comment]
Yes any client can perform a read request. One could argue that maybe the client would still need to be "registered" (i.e. have an open stream), but I don't think there is a reason to require a client to do the arbitration handshake if it simply needs to perform a read.
